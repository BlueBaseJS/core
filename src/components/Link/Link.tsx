import { LinkProps, NavigationActionsObject } from '@bluebase/components';
import { NativeSyntheticEvent, NativeTouchEvent, Platform } from 'react-native';

import { NavigationActions } from '../../getComponent';
import React from 'react';
import { TouchableItem } from '../TouchableItem';

/**
 * ðŸ”— Link
 */
export class Link extends React.PureComponent<LinkProps> {
	public static defaultProps: Partial<LinkProps> = {
		component: TouchableItem,
		replace: false,
	};

	handlePress(event: NativeSyntheticEvent<NativeTouchEvent>, navigation: NavigationActionsObject) {
		const { method, routeName, path, params, replace } = this.props;

		if (event && event.defaultPrevented) {
			event.preventDefault();

			let fn = navigation.navigate;

			if (method) {
				fn = navigation[method];
			}

			if (replace === true) {
				fn = navigation.replace;
			}

			if (typeof routeName === 'string') {
				fn(routeName, params);
			} else if (typeof path === 'string') {
				fn({ path }, params);
			}
		}
	}

	render() {
		const { component: Component, routeName, path, params, onPress, replace, ...rest } = this.props;

		return Component ? (
			<NavigationActions>
				{(navigation: NavigationActionsObject) => {
					const onPressDefault = (e: any) => this.handlePress(e, navigation);

					return Platform.OS === 'web' && !!path && path !== '' ? (
						<a href={path} onClick={(onPress as any) || onPressDefault} {...this.props} />
					) : (
							<Component {...rest} onPress={onPress || onPressDefault} />
						);
				}}
			</NavigationActions>
		) : null;
	}
}
