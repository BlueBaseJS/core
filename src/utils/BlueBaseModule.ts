import { MaybeEsModule, getDefiniteModule } from './Modules';
import { MaybePromise, getDefinitePromise, isPromise } from './Promises';

/**
 * Structure of an input acceptable to create a BlueBaseModule
 */
// export type BlueBaseModuleInput<T> = MaybePromise<MaybeEsModule<T>>;
export type BlueBaseModuleInput<T> = MaybeEsModule<MaybePromise<MaybeEsModule<T>>>;

/**
 * An object that may or may not be an instance of BlueBaseModule
 */
export type MaybeBlueBaseModule<T> = T | BlueBaseModule<T>;

/**
 * An object that maybe an instance of BlueBaseModule, or a BlueBaseModuleInput object, or neiter.
 */
export type MaybeBlueBaseModuleOrInput<T> = BlueBaseModuleInput<T> | MaybeBlueBaseModule<T>;

/**
 * Checks if an input object is a BlueBaseModule. Returns as is if true, otherwise,
 * wraps the input into a BlueBaseModule and returns.
 *
 * @param input Input object
 */
export function getDefiniteBlueBaseModule<T>(input: MaybeBlueBaseModuleOrInput<T>): BlueBaseModule<T> {

	if (isBlueBaseModule(input)) {
		return input;
	}

	return new BlueBaseModule(input) as BlueBaseModule<T>;
}

/**
 * Check if an input object is an instance of BlueBaseModule
 * @param input Input object
 */
export function isBlueBaseModule<T>(input: MaybeBlueBaseModuleOrInput<T>): input is BlueBaseModule<T> {
	return (input instanceof BlueBaseModule);
}

/**
 * ðŸ“¦ BlueBase Module
 */
export class BlueBaseModule<T> {

	/**
	 * This is the input value. If this value is an ES module,
	 * then we store the value of `.default` property.
	 *
	 * This element may (in case of code splitting) or may not
	 * be a Promise. Also, the promise itself may or may not
	 * resolve an ES module.
	 */
	public module: MaybePromise<MaybeEsModule<T>>;

	/**
	 * The module is converted into promise. This promise
	 * resolves the module itself. If the module is an ES module,
	 * then the promise resolve `.default` property of th module.
	 *
	 * This means, this promise will always resolve the intended
	 * module, no matter if it was a promise or not, or an ES module
	 * or not.
	 */
	public promise: Promise<T>;

	/**
	 * This flag tells if the initial input module as a promise.
	 * If it is false, then it means that the module property
	 * represents the actual module, converted from ES.
	 */
	public isAsync: boolean = false;

	constructor(input: BlueBaseModuleInput<T>) {

		this.module = getDefiniteModule(input);

		if (isPromise(this.module)) {
			this.isAsync = true;
		}

		this.promise = new Promise((resolve, reject) => {
			getDefinitePromise(this.module)
				.then((m: MaybeEsModule<T>) => resolve(getDefiniteModule(m)))
				.catch(reject);
		});
	}
}
