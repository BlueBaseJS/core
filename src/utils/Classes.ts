/**
 * Checks if a variable is a class
 * @param input Input object to test
 */
export function isClass(input: any) {
	return typeof input === 'function'
		&& /^class\s/.test(Function.prototype.toString.call(input));
}
