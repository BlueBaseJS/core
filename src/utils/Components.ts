export type RenderPropChildren<Props = any> = ((props: Props) => React.ReactNode);

export type MaybeRenderPropChildren<T = any> = RenderPropChildren<T> | React.ReactNode;

export function renderChildrenWithProps<T = any>(children: MaybeRenderPropChildren<T>, props: T) {
	if (typeof children === 'function') {
		return (children as any)(props);
	}

	return children;
}