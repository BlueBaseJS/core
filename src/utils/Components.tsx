import React from 'react';

export type RenderPropChildren<Props = any> = (props: Props) => React.ReactNode;

export type MaybeRenderPropChildren<T = any> = RenderPropChildren<T> | React.ReactNode;

export function renderChildrenWithProps<T = any>(children: MaybeRenderPropChildren<T>, props: T) {
	if (typeof children === 'function') {
		return (children as any)(props);
	}

	return children || null;
}

/**
 * Use this HOC on the raw component, before registering in BlueBase
 * @param Component
 */
export function applyRef<T = any>(Component: React.ComponentType<T>): React.ComponentType<T> {
	return ({ bluebaseRef, ...rest }: any) => <Component ref={bluebaseRef} {...rest} />;
}

/**
 * Use this HOC on the component resolved from BlueBase or "getComponent" method
 * @param Component
 */
export function forwardRef<T = any>(Component: React.ComponentType<T>): React.ComponentType<T> {
	const ForwardRefComponent = Component as any;
	return React.forwardRef((props, ref) => {
		return <ForwardRefComponent {...props} bluebaseRef={ref} />;
	}) as any;
}
