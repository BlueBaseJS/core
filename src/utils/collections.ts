/**
 * An object that may or may not be an array
 */
export type MaybeArray<T> = T | T[];

/**
 * If an input is an array, returns as is. Otherwise creates a new array with input object at index 0 and returns it.
 * @param input Input object
 */
export function getDefiniteArray<T>(input: MaybeArray<T>): T[] {
	return Array.isArray(input) ? input : [input];
}