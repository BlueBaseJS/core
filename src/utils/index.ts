export * from './BlueBaseModule';
export * from './Collections';
export * from './Components';
export * from './Misc';
export * from './Modules';
export * from './Paths';
export * from './Promises';
export * from './Thunks';
export * from './Typescript';
