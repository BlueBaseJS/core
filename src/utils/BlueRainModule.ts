import { MaybeEsModule, getDefiniteModule, isEsModule } from '../utils/Modules';
import { MaybePromise, getDefinitePromise, isPromise } from '../utils/Promises';

/**
 * Structure of an input acceptable to create a BlueRainModule
 */
// export type BlueRainModuleInput<T> = MaybePromise<MaybeEsModule<T>>;
export type BlueRainModuleInput<T> = MaybeEsModule<MaybePromise<MaybeEsModule<T>>>;

/**
 * An object that may or may not be an instance of BlueRainModule
 */
export type MaybeBlueRainModule<T> = T | BlueRainModule<T>;

/**
 * An object that maybe an instance of BlueRainModule, or a BlueRainModuleInput object, or neiter.
 */
export type MaybeBlueRainModuleOrInput<T> = BlueRainModuleInput<T> | MaybeBlueRainModule<T>;

/**
 * Checks if an input object is a BlueRainModule. Returns as is if true, otherwise,
 * wraps the input into a BlueRainModule and returns.
 *
 * @param input Input object
 */
export function getDefiniteBlueRainModule<T>(input: MaybeBlueRainModuleOrInput<T>): BlueRainModule<T> {

	if (isBlueRainModule(input)) {
		return input;
	}

	return new BlueRainModule(input) as BlueRainModule<T>;
}

/**
 * Check if an input object is an instance of BlueRainModule
 * @param input Input object
 */
export function isBlueRainModule<T>(input: MaybeBlueRainModuleOrInput<T>): input is BlueRainModule<T> {
	return (input instanceof BlueRainModule);
}

/**
 * ðŸ“¦ BlueRain Module
 */
export class BlueRainModule<T> {

	/**
	 * This is the input value. If this value is an ES module,
	 * then we store the value of `.default` property.
	 *
	 * This element may (in case of code splitting) or may not
	 * be a Promise. Also, the promise itself may or may not
	 * resolve an ES module.
	 */
	public module: MaybePromise<MaybeEsModule<T>>;

	/**
	 * The module is converted into promise. This promise
	 * resolves the module itself. If the module is an ES module,
	 * then the promise resolve `.default` property of th module.
	 *
	 * This means, this promise will always resolve the intended
	 * module, no matter if it was a promise or not, or an ES module
	 * or not.
	 */
	public promise: Promise<T>;

	/**
	 * This flag tells if the initial input module as a promise.
	 * If it is false, then it means that the module property
	 * represents the actual module, converted from ES.
	 */
	public isAsync: boolean = false;

	constructor(input: BlueRainModuleInput<T> ) {

		this.module = (isEsModule(input)) ? getDefiniteModule(input) : input;

		if (isPromise(this.module)) {
			this.isAsync = true;
		}

		this.promise = new Promise((resolve, reject) => {
			getDefinitePromise(this.module)
				.then((m: MaybeEsModule<T>) => resolve(getDefiniteModule(m)))
				.catch(reject);
		});
	}
}
