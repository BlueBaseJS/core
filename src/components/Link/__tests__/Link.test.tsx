import { BlueBaseApp } from '../../../';
import { Link } from '../Link';
import React from 'react';
import { mount } from 'enzyme';
import { renderChildrenWithProps } from '../../../utils';
import { waitForElement } from 'enzyme-async-helpers';

let stubActions: any = {};

export const NavigationActions = ({ children }: any) =>
	renderChildrenWithProps(children, stubActions);

describe('Link', () => {
	beforeEach(() => {
		stubActions = {
			getParam: jest.fn(),
			goBack: jest.fn(),
			navigate: jest.fn(),
			pop: jest.fn(),
			push: jest.fn(),
			replace: jest.fn(),
			setParams: jest.fn(),
			source: null,
			state: {
				key: '',
				params: {},
				routeName: '',
				url: '',
			},
		};
	});

	test(`should call the navigate function with the given routeName`, async () => {
		const components = {
			NavigationActions,
		};

		const wrapper = mount(
			<BlueBaseApp components={components}>
				<Link routeName="Foo" />
			</BlueBaseApp>
		);

		// Wait for render
		await waitForElement(wrapper as any, Link);

		const onPress: any = wrapper.find('Link TouchableItem').prop('onPress');
		onPress(
			{
				defaultPrevented: false,
				preventDefault: () => {
					return;
				},
			},
			stubActions
		);

		expect(stubActions.navigate).toBeCalledTimes(1);
		expect(stubActions.navigate).toBeCalledWith('Foo', undefined);

		expect(stubActions.push).toBeCalledTimes(0);
		expect(stubActions.replace).toBeCalledTimes(0);
	});

	test(`should call the push function with the given routeName`, async () => {
		const components = {
			NavigationActions,
		};

		const wrapper = mount(
			<BlueBaseApp components={components}>
				<Link routeName="Foo" method="push" />
			</BlueBaseApp>
		);

		// Wait for render
		await waitForElement(wrapper as any, Link);

		const onPress: any = wrapper.find('Link TouchableItem').prop('onPress');
		onPress(
			{
				defaultPrevented: false,
				preventDefault: () => {
					return;
				},
			},
			stubActions
		);

		expect(stubActions.push).toBeCalledTimes(1);
		expect(stubActions.push).toBeCalledWith('Foo', undefined);

		expect(stubActions.navigate).toBeCalledTimes(0);
		expect(stubActions.replace).toBeCalledTimes(0);
	});

	test(`should call the push function with the given path`, async () => {
		jest.mock('Platform', () => {
			const Platform = (require as any).requireActual('Platform');
			Platform.OS = 'web';
			return Platform;
		});

		const components = {
			NavigationActions,
		};

		const wrapper = mount(
			<BlueBaseApp components={components}>
				<Link path="/foo" method="push" />
			</BlueBaseApp>
		);

		// Wait for render
		await waitForElement(wrapper as any, Link);

		const onPress: any = wrapper.find('Link a').prop('onClick');
		onPress(
			{
				defaultPrevented: false,
				preventDefault: () => {
					return;
				},
			},
			stubActions
		);

		expect(stubActions.push).toBeCalledTimes(1);
		expect(stubActions.push).toBeCalledWith({ path: '/foo' }, undefined);

		expect(stubActions.navigate).toBeCalledTimes(0);
		expect(stubActions.replace).toBeCalledTimes(0);
	});

	test(`should call the replace function with the given path`, async () => {
		const components = {
			NavigationActions,
		};

		const wrapper = mount(
			<BlueBaseApp components={components}>
				<Link path="/foo" replace />
			</BlueBaseApp>
		);

		// Wait for render
		await waitForElement(wrapper as any, Link);

		const onPress: any = wrapper.find('Link a').prop('onClick');
		onPress(
			{
				defaultPrevented: false,
				preventDefault: () => {
					return;
				},
			},
			stubActions
		);

		expect(stubActions.replace).toBeCalledTimes(1);
		expect(stubActions.replace).toBeCalledWith({ path: '/foo' }, undefined);

		expect(stubActions.push).toBeCalledTimes(0);
	});

	test(`should not do anything if event.defaultPrevented is true`, async () => {
		const components = {
			NavigationActions,
		};

		const wrapper = mount(
			<BlueBaseApp components={components}>
				<Link path="/foo" replace />
			</BlueBaseApp>
		);

		// Wait for render
		await waitForElement(wrapper as any, Link);

		const onPress: any = wrapper.find('Link a').prop('onClick');
		onPress(
			{
				defaultPrevented: true,
				preventDefault: () => {
					return;
				},
			},
			stubActions
		);

		expect(stubActions.replace).toBeCalledTimes(0);
		expect(stubActions.push).toBeCalledTimes(0);
	});

	test(`should call the push function with the given path`, async () => {
		const customOnPress = jest.fn();

		const wrapper = mount(
			<BlueBaseApp>
				<Link path="/foo" onPress={customOnPress} />
			</BlueBaseApp>
		);

		// Wait for render
		await waitForElement(wrapper as any, Link);

		const onPress: any = wrapper.find('Link a').prop('onClick');
		onPress(
			{
				defaultPrevented: false,
				preventDefault: () => {
					return;
				},
			},
			stubActions
		);

		expect(customOnPress).toBeCalledTimes(1);

		expect(stubActions.push).toBeCalledTimes(0);
		expect(stubActions.replace).toBeCalledTimes(0);
	});

	test(`should not do anything if there is no path or routeName prop`, async () => {
		const wrapper = mount(
			<BlueBaseApp>
				<Link />
			</BlueBaseApp>
		);

		// Wait for render
		await waitForElement(wrapper as any, Link);

		const onPress: any = wrapper.find('Link TouchableItem').prop('onPress');
		onPress(
			{
				defaultPrevented: false,
				preventDefault: () => {
					return;
				},
			},
			stubActions
		);

		expect(stubActions.push).toBeCalledTimes(0);
		expect(stubActions.replace).toBeCalledTimes(0);
	});

	test(`should not render anything if component prop is null`, async () => {
		const wrapper = mount(
			<BlueBaseApp>
				<Link component={null as any} />
			</BlueBaseApp>
		);

		// Wait for render
		await waitForElement(wrapper as any, Link);

		expect(
			wrapper
				.find('Link')
				.last()
				.children().length
		).toBe(0);
	});
});
