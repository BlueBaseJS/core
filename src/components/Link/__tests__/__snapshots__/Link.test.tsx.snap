// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Link should call the push function with the given path 1`] = `
Object {
  "getParam": [MockFunction],
  "goBack": [MockFunction],
  "navigate": [MockFunction],
  "pop": [MockFunction],
  "push": [MockFunction] {
    "calls": Array [
      Array [
        Object {
          "path": "/foo",
        },
        undefined,
      ],
    ],
    "results": Array [
      Object {
        "type": "return",
        "value": undefined,
      },
    ],
  },
  "replace": [MockFunction],
  "setParams": [MockFunction],
  "source": null,
  "state": Object {
    "key": "",
    "params": Object {},
    "routeName": "",
    "url": "",
  },
}
`;

exports[`Link should call the push function with the given path 2`] = `
Object {
  "getParam": [MockFunction],
  "goBack": [MockFunction],
  "navigate": [MockFunction],
  "pop": [MockFunction],
  "push": [MockFunction],
  "replace": [MockFunction],
  "setParams": [MockFunction],
  "source": null,
  "state": Object {
    "key": "",
    "params": Object {},
    "routeName": "",
    "url": "",
  },
}
`;

exports[`Link should call the push function with the given routeName 1`] = `
Object {
  "getParam": [MockFunction],
  "goBack": [MockFunction],
  "navigate": [MockFunction],
  "pop": [MockFunction],
  "push": [MockFunction] {
    "calls": Array [
      Array [
        "Foo",
        undefined,
      ],
    ],
    "results": Array [
      Object {
        "type": "return",
        "value": undefined,
      },
    ],
  },
  "replace": [MockFunction],
  "setParams": [MockFunction],
  "source": null,
  "state": Object {
    "key": "",
    "params": Object {},
    "routeName": "",
    "url": "",
  },
}
`;

exports[`Link should call the replace function with the given path 1`] = `
Object {
  "getParam": [MockFunction],
  "goBack": [MockFunction],
  "navigate": [MockFunction],
  "pop": [MockFunction],
  "push": [MockFunction],
  "replace": [MockFunction] {
    "calls": Array [
      Array [
        Object {
          "path": "/foo",
        },
        undefined,
      ],
    ],
    "results": Array [
      Object {
        "type": "return",
        "value": undefined,
      },
    ],
  },
  "setParams": [MockFunction],
  "source": null,
  "state": Object {
    "key": "",
    "params": Object {},
    "routeName": "",
    "url": "",
  },
}
`;

exports[`Link should not do anything if event.defaultPrevented is true 1`] = `
Object {
  "getParam": [MockFunction],
  "goBack": [MockFunction],
  "navigate": [MockFunction],
  "pop": [MockFunction],
  "push": [MockFunction],
  "replace": [MockFunction],
  "setParams": [MockFunction],
  "source": null,
  "state": Object {
    "key": "",
    "params": Object {},
    "routeName": "",
    "url": "",
  },
}
`;

exports[`Link should not do anything if there is no path or routeName prop 1`] = `
Object {
  "getParam": [MockFunction],
  "goBack": [MockFunction],
  "navigate": [MockFunction],
  "pop": [MockFunction],
  "push": [MockFunction],
  "replace": [MockFunction],
  "setParams": [MockFunction],
  "source": null,
  "state": Object {
    "key": "",
    "params": Object {},
    "routeName": "",
    "url": "",
  },
}
`;
